lib/Twig/Extension/Core.php

```diff
@@ -1,14 +1,1726 @@
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
 
-class_exists('Twig\Extension\CoreExtension');
+namespace Twig\Extension {
+use Twig\ExpressionParser;
+use Twig\Node\Expression\Binary\AddBinary;
+use Twig\Node\Expression\Binary\AndBinary;
+use Twig\Node\Expression\Binary\BitwiseAndBinary;
+use Twig\Node\Expression\Binary\BitwiseOrBinary;
+use Twig\Node\Expression\Binary\BitwiseXorBinary;
+use Twig\Node\Expression\Binary\ConcatBinary;
+use Twig\Node\Expression\Binary\DivBinary;
+use Twig\Node\Expression\Binary\EndsWithBinary;
+use Twig\Node\Expression\Binary\EqualBinary;
+use Twig\Node\Expression\Binary\FloorDivBinary;
+use Twig\Node\Expression\Binary\GreaterBinary;
+use Twig\Node\Expression\Binary\GreaterEqualBinary;
+use Twig\Node\Expression\Binary\InBinary;
+use Twig\Node\Expression\Binary\LessBinary;
+use Twig\Node\Expression\Binary\LessEqualBinary;
+use Twig\Node\Expression\Binary\MatchesBinary;
+use Twig\Node\Expression\Binary\ModBinary;
+use Twig\Node\Expression\Binary\MulBinary;
+use Twig\Node\Expression\Binary\NotEqualBinary;
+use Twig\Node\Expression\Binary\NotInBinary;
+use Twig\Node\Expression\Binary\OrBinary;
+use Twig\Node\Expression\Binary\PowerBinary;
+use Twig\Node\Expression\Binary\RangeBinary;
+use Twig\Node\Expression\Binary\StartsWithBinary;
+use Twig\Node\Expression\Binary\SubBinary;
+use Twig\Node\Expression\Filter\DefaultFilter;
+use Twig\Node\Expression\NullCoalesceExpression;
+use Twig\Node\Expression\EmptyCoalesceExpression
+use Twig\Node\Expression\Test\ConstantTest;
+use Twig\Node\Expression\Test\DefinedTest;
+use Twig\Node\Expression\Test\DivisiblebyTest;
+use Twig\Node\Expression\Test\EvenTest;
+use Twig\Node\Expression\Test\NullTest;
+use Twig\Node\Expression\Test\OddTest;
+use Twig\Node\Expression\Test\SameasTest;
+use Twig\Node\Expression\Unary\NegUnary;
+use Twig\Node\Expression\Unary\NotUnary;
+use Twig\Node\Expression\Unary\PosUnary;
+use Twig\TokenParser\BlockTokenParser;
+use Twig\TokenParser\DeprecatedTokenParser;
+use Twig\TokenParser\DoTokenParser;
+use Twig\TokenParser\EmbedTokenParser;
+use Twig\TokenParser\ExtendsTokenParser;
+use Twig\TokenParser\FilterTokenParser;
+use Twig\TokenParser\FlushTokenParser;
+use Twig\TokenParser\ForTokenParser;
+use Twig\TokenParser\FromTokenParser;
+use Twig\TokenParser\IfTokenParser;
+use Twig\TokenParser\ImportTokenParser;
+use Twig\TokenParser\IncludeTokenParser;
+use Twig\TokenParser\MacroTokenParser;
+use Twig\TokenParser\SetTokenParser;
+use Twig\TokenParser\SpacelessTokenParser;
+use Twig\TokenParser\UseTokenParser;
+use Twig\TokenParser\WithTokenParser;
+use Twig\TwigFilter;
+use Twig\TwigFunction;
+use Twig\TwigTest;
 
-@trigger_error(sprintf('Using the "Twig_Extension_Core" class is deprecated since Twig version 2.7, use "Twig\Extension\CoreExtension" instead.'), E_USER_DEPRECATED);
+final class CoreExtension extends AbstractExtension
```

Comment:
this change is wrong

Reply:
Correct, the error message is wrong (a copy and paste error) but if we want this removed as per your earlier comment, then it probably doesn't matter. It'd be fixed by just removing the whole deprecation nonsense.

Reply:
I pushed a change to remove the pointless `Twig_Node_Expression_EmptyCoalesce` class entirely, which should resolve this.

Reply:
this is not fixed. You are still making a huge mess in the CoreExtension file, reverting the legacy file to a full class.

Reply:
I'm using this to merge into: https://github.com/twigphp/Twig/blob/2.x/src/Extension/CoreExtension.php

Should I be using it from a different branch?

Reply:
@stof I'd love to get this addressed, but I could use some guidance.

The only reason there is a change to `lib/Twig/Extension/Core.php` is to inclde the `???` operator. To fix this so that it's not creating a "huge mess" I just need to know what branch/source I should be diff'ing with to add this line?

Reply:
you should not be making any change in `lib/Twig/Extension/Core.php`, which is a legacy BC file (except you are reverting this to not be a BC file anymore). The change should be done only in `src/Extension/CoreExtension.php`

Reply:
your PR is reverting #2863 on this file, which is what the huge mess is.

Reply:
I see. That change to make `lib/Twig/Extension/Core.php` wasn't made until after the PR I made, thus my confusion about what's going on here.

I will revert the now legacy `lib/Twig/Extension/Core.php` and make my changes in `src/Extension/CoreExtension.php`